keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,997 @@
+                        if !pluginStatus.IsUnschedulable() {
+```
+
+##### Resulting output
+
+Here is log output from `kube-scheduler -v5` for a Pod with an inline volume which cannot be created because storage is exhausted:
+
+```
+I1026 16:21:00.461394  801139 scheduler.go:436] "Attempting to schedule pod" pod="default/my-csi-app-inline-volume"
+I1026 16:21:00.461476  801139 binder.go:730] PreFilter/VolumeBinding: "PVC is not bound" pod="default/my-csi-app-inline-volume" pvc="default/my-csi-app-inline-volume-my-csi-volume"
+```
+
+The next line is from a file which has not been converted. It’s not clear in which context that message gets emitted:
+````
+I1026 16:21:00.461619  801139 csi.go:222] "Persistent volume had no name for claim" PVC="default/my-csi-app-inline-volume-my-csi-volume"
+````
+
+```
+I1026 16:21:00.461647  801139 binder.go:266] NominatedPods/Filter/VolumeBinding: "FindPodVolumes starts" pod="default/my-csi-app-inline-volume" node="127.0.0.1"
+I1026 16:21:00.461673  801139 binder.go:842] NominatedPods/Filter/VolumeBinding: "No matching volumes for PVC on node" pod="default/my-csi-app-inline-volume" node="127.0.0.1" default/my-csi-app-inline-volume-my-csi-volume="127.0.0.1"
+I1026 16:21:00.461724  801139 binder.go:971] NominatedPods/Filter/VolumeBinding: "Node has no accessible CSIStorageCapacity with enough capacity" pod="default/my-csi-app-inline-volume" node="127.0.0.1" pvc="default/my-csi-app-inline-volume-my-csi-volume" pvcSize=549755813888000 sc="csi-hostpath-fast"
+I1026 16:21:00.461817  801139 preemption.go:195] "Preemption will not help schedule pod on any node" pod="default/my-csi-app-inline-volume"
+I1026 16:21:00.461886  801139 scheduler.go:464] "Status after running PostFilter plugins for pod" pod="default/my-csi-app-inline-volume" status=&{code:2 reasons:[0/1 nodes are available: 1 Preemption is not helpful for scheduling.] err:<nil> failedPlugin:}
+I1026 16:21:00.461918  801139 factory.go:209] "Unable to schedule pod; no fit; waiting" pod="default/my-csi-app-inline-volume" err="0/1 nodes are available: 1 node(s) did not have enough free storage."
+```
+
+### Test Plan
+
+The new code will be covered by unit tests that execute as part of `pull-kubernetes-unit`.
+
+Converted components will be tested by exercising them with JSON output at high
+log levels to emit as many log messages as possible. Analysis of those logs
+will detect duplicate keys that might occur when a caller uses `WithValues` and
+a callee adds the same value in a log message.
+
+Static code analysis cannot detect this, or at least not easily. What it can
+check is that the individual log calls pass valid key/value pairs.
+
+
+### Graduation Criteria
+
+#### Alpha
+
+- Common utility code available (logcheck with the additional checks,
+  `k8s.io/klogr` v0.1.0)
+- Documentation for developers available at https://github.com/kubernetes/community
+- At least kube-scheduler framework and some plugins converted
```

Comment:
We have ack from aldo so :+1: 

Which plugins?

Reply:
For the performance test I would say "all of them", but then getting all of those changes merged may take longer because the initial conversion can be quick-and-dirty while the actual conversion must choose keys for structured logging calls more carefully and the result has to be reviewed by different SIGs.

So I'm a bit undecided what we actually want to put into the KEP.

Reply:
Maybe just give some possible examples that we don't have to commit to... I wasn't sure what you meant here by "plugins", it's very vague

Reply:
I made it more specific and explicitly called out volumebinding and nodevolumelimits because they make the most use of klog:
```
$ for i in pkg/scheduler/framework/plugins/*; do echo "$(git grep klog $i | wc -l) $i"; done | sort -n
0 pkg/scheduler/framework/plugins/feature
0 pkg/scheduler/framework/plugins/helper
0 pkg/scheduler/framework/plugins/imagelocality
0 pkg/scheduler/framework/plugins/names
0 pkg/scheduler/framework/plugins/nodeaffinity
0 pkg/scheduler/framework/plugins/nodename
0 pkg/scheduler/framework/plugins/nodeports
0 pkg/scheduler/framework/plugins/nodeunschedulable
0 pkg/scheduler/framework/plugins/queuesort
0 pkg/scheduler/framework/plugins/README.md
0 pkg/scheduler/framework/plugins/registry.go
0 pkg/scheduler/framework/plugins/selectorspread
0 pkg/scheduler/framework/plugins/tainttoleration
0 pkg/scheduler/framework/plugins/testing
0 pkg/scheduler/framework/plugins/volumerestrictions
2 pkg/scheduler/framework/plugins/defaultbinder
2 pkg/scheduler/framework/plugins/examples
3 pkg/scheduler/framework/plugins/defaultpreemption
3 pkg/scheduler/framework/plugins/volumezone
4 pkg/scheduler/framework/plugins/noderesources
5 pkg/scheduler/framework/plugins/interpodaffinity
5 pkg/scheduler/framework/plugins/podtopologyspread
22 pkg/scheduler/framework/plugins/nodevolumelimits
58 pkg/scheduler/framework/plugins/volumebinding
```